### PR TITLE
3D documentation addition

### PR DESCRIPTION
**Pull request by Hwurzburg**
Added notes on how to setup 3D mode

### CODE REVIEW

**Comment by dkisselev** [comment]
>  set esc max pwm in the ESC GUI to the maximum your transmitter outputs on the throttle channel (normally 2000us)

Are you sure about this? I'm pretty sure the naze doesn't directly passthrough any RC channels to the ESC's, they all go through PID's. So the transmitter outputs shouldn't matter, it's what the FC puts out. Later in the guide you set maxthrottle and mincommand to 2000 and 1000, so that's what the ESC's should listen to (as in, your instructions will still work, they're just not 100% factually correct).

**Comment by Hwurzburg** [comment]
yes, as I read the code (and tested on the bench)...RC throttle is scaled by maxthrottle , midthrottle and expo(if expo is nonzero) and serves as the base value to all motors...PIDS are added, and output values constrained to either minthrottle/maxthrottle (non 3D) or mincommand/maxthrottle (3D), and also clipping prevention is added to assure no individual motor saturates (allows full stablization at full throttle)

**Comment by dkisselev** [comment]
Right, so the constraints are mincommand/maxthrottle as configured in the flight controller. So that's what the ESC's should base their output power on.

The ```transmitter outputs on the throttle channel ```, as the current documentation states are not relevant because the FC will constrain them to the above parameters no matter what.

**Comment by Hwurzburg** [comment]
correct, so putting maxthrottle at 1850 when full stick is 2000 creates a deadzone on the stick from where the TX outputs 1850 to where it outputs 2000....stick movement in that range results in no change in FC output....same on low end...very undesirable...so to get FC output for all of the throttle stick range you have to configure as I stated

**Comment by dkisselev** [comment]
Oh, I see what you're saying, you're saying calibrate the ESC's based on the FC output values when the TX is sending out both full and no throttle? That way you can see the full range that the FC puts out.

In that case we will probably want to clarify that what you're looking to calibrate against is the output values. Manually setting the limits is also not really the recommended way of doing it. Typically you want to put the ESC's in to calibration mode, then either throttle up to 100/0, or use the motors tab in the configurator to get the FC to put out the 100% and 0% PWM values.

**Comment by Hwurzburg** [comment]
not quite...your are setting it up so that the FC outputs the correct values matching your TX stick range and setting the ESCs to match this also...since you HAVE to use the ESC GUI to even get into 3D mode, its easier to cal the esc there....the normal esc cal mode is used since most do not use the ESC gui to setup the ESC...you have to do that for 3D...this is quick and works...I did add a note about how to find the TX throttle output values you are setting and that are referred to constantly in setup...

**Comment by SystemsGuy** [comment]
Excellent writeup!

**Comment by thenickdude** [comment]
>correct, but you miss the point.... putting maxthrottle at 1850 when full stick is 2000 creates a deadzone on the stick from where the TX outputs 1850 to where it outputs 2000....stick movement in that range results in no change in FC output

That isn't true, the input throttle value range from the Rx [1000...2000] is rescaled by CF to fill the range [minthrottle...maxthrottle] and that becomes `rcCommand[THROTTLE]` which is used to drive the PID controller. 

My throttle stick ranges from 1000 to 2000. I set these settings to constrain the motor outputs in the tiny range [1100...1200]:

```
feature 3d
set min_command = 1100
set max_throttle = 1200
set 3d_deadband_low = 1140
set 3d_deadband_high = 1160
set 3d_neutral = 1150
```

Try it on the Motors tab in the configurator (no props). The motor commands move smoothly throughout the entire range of throttle stick travel, there's absolutely no dead zone.

**Comment by Hwurzburg** [comment]
The motors tab does not use the RC input....it uses the config slider....no correlation to rc input range received.... it always runs from Mincommand to Maxthrottle....so you cant test it that way...
I may be reading the code wrong, but in mw.c:
"tmp = constrain(rcData[THROTTLE], masterConfig.rxConfig.mincheck, PWM_RANGE_MAX);
tmp = (uint32_t)(tmp - masterConfig.rxConfig.mincheck) * PWM_RANGE_MIN / (PWM_RANGE_MAX - masterConfig.rxConfig.mincheck);
 tmp2 = tmp / 100;
    rcCommand[THROTTLE] = lookupThrottleRC[tmp2] + (tmp - tmp2 * 100) * (lookupThrottleRC[tmp2 + 1] - lookupThrottleRC[tmp2]) / 100;"
takes the RC data, constrains it to MINCHECK->2000, scales to 0-1000,then looks up the segment value of a curve set by MIDTHROTTLE,EXPO,MINTHROTTLE, and MAXTHROTTLE from a table set at init and interpolates between the 12 points of that curve:
"lookupThrottleRC[i] = 10 * controlRateConfig->thrMid8 + tmp * (100 - controlRateConfig->thrExpo8 + (int32_t) controlRateConfig->thrExpo8 * (tmp * tmp) / (y * y)) / 10;
        lookupThrottleRC[i] = escAndServoConfig->minthrottle + (int32_t) (escAndServoConfig->maxthrottle - escAndServoConfig->minthrottle) * lookupThrottleRC[i] / 1000;"
so with an rc data input of 2000 and MAXTHROTTLE at 1850, it will fetch the last value (of 12) in the table...which had been output also for all rc inputs above 1850 which is the last segment set in the lookup table...ie a flat region of the same actual throttle for varying input...ie deadband...
this has correlated to what I see on the bench actually running the motors from the RC input (not the sliders tab, which is different)....admittedly I didnt check it more than once nor extensively...but also seems to correlate to what I see in the air

**Comment by thenickdude** [comment]
Yeah I meant to observe the motor values being transmitted on the motors
tab, not to use the sliders to set the motors.

**Comment by Hwurzburg** [comment]
after running some tests, you are correct....the code does scale pwm input =2000 to be MAXTHROTTLE..(which in and of itself has some interesting implications if you dont cal the escs via the RC and not the motors tab...think of the case when RC TX puts out 1850us max at full stick but MAXTHROTTLE is set to 1850 and escs cal'd by motors tab....max to motor will only be ~ 1850/2000 of full power that the esc could output)
this is why I thought I saw a "deadband" on one TX setup....it had max 1900us out...
anyway, so the doc's setup method is correct, even if my explanation to dkisselev was incorrect...

File: docs/3d.md [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+
+ **3d_deadband_low**: (typically 1450) lowest value for inverted throttle output when armed, highest is mincommand 
+
+                 (Note: switchover point for upright or inverted is **midrc** on throttle stick)
+
+ **3d_neutral**: (typically 1500) same effect as mincommand, but for 3d operation...ie disarmed pulsewidth to escs 
+
+ **3d_deadband_throttle**: (typically 50) throttle stick range around midrc for allowing arming to occur 
+
+ **maxthrottle**= max tx output (typically 2000us)
```
Comment (thenickdude) [inline docs/3d.md]: Max signal sent to ESCs (no reference to Tx)

File: docs/3d.md [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+
+ **mincommand** = min tx output (typically 1000us)
```
Comment (thenickdude) [inline docs/3d.md]: Minimum signal sent to ESCs (no reference to Tx)

File: docs/3d.md [hunk 3]
```diff
@@ -0,0 +1,36 @@
+#3D Setup
+
+When FEATURE 3D is enabled,either by the configuration GUI or CLI command, it allows operation with bi-directional escs to provide inverted as well as normal,upright operation. 
+
+##Setup:
+
+* Enable **FEATURE 3D** in the firmware
+
+* Setup ESC to **BIDIRECTIONAL** (norm/reverse/birectional slider in BLHeli GUI) and set **esc max pwm** in the ESC GUI to the maximum your transmitter outputs on the throttle channel (normally 2000us) and **esc min pwm** to the minimum (normally 1000us) and the **midpoint** to halfway between (normally 1500us). You can determine what your TX outputs by using the **Reciever** tab in the configurator.
```
Comment (thenickdude) [inline docs/3d.md]: Should be no reference to the Tx in this portion.

**Comment by Hwurzburg** [comment]
Sorry, Nick, but I disagree with your comments on the doc....the methodology, if followed USING the actual TX throttle outputs as shown in the RX tab, and then tailoring the MAXTHROTTLE , MINCOMMAND values to these and cal'ing the esc settings (you HAVE to use BLHeli to make 3d work) results in maximum resolution vs stick, and is easiest for a newbie to follow and get success at 3d setup...you are forcing the FC to follow the TX, so it should be referenced IMO

**Comment by thenickdude** [comment]
Maximum resolution is achieved when mincommand/maxthrottle has the greatest range the ESCs can distinguish, it doesn't have anything to do with the Tx throttle stick. Note that most of the command sent to the ESCs is determined by CF's PID controller, not directly from the Tx stick, so there is no benefit in crippling the resolution to match the Tx (e.g. if the Tx's throttle stick range was 1000...1500, but the ESCs could distinguish 2000 steps, setting max_throttle to 1500 would halve the resolution that CF has available for maintaining attitude. Yes, you wouldn't lose any resolution from your Tx stick, but CF's control over the rotation of the craft will halve in resolution).

**Comment by Hwurzburg** [comment]
Edit as you see fit....

**Comment by hydra** [comment]
Hi guys.  Could someone either a) update this PR or b) create a new one that builds on this one with the additional corrections/fixes?

@thenickdude @Hwurzburg ?

**Comment by hydra** [comment]
@Hwurzburg update?

**Pull request closed without merging**